### ISSUE TITLE
running some models using webgl is 10x slower than using nodejs

### ISSUE THREAD

**Issue by vladmandic** [tensorflow/tfjs#4871]
Model executes in NodeJS using `tensorflow` backend in ~100ms, but above 1sec in browser using `WebGL` backend
That is 10x difference in performance between NodeJS and `WebGL`  

I've tried to figure out why a model is executing an order of magnitude slower than expected using `WebGL` backend,  
but profiler info for it makes no sense

```
const t0 = performance.now();
const res = await tf.profile(() => model.executeAsync(input));
const t1 = performance.now();
const wallTime = t1 - t0;
const kernelTime = res.kernels.reduce((a, b) => a += b.kernelTimeMs, 0);
```

wallTime is 900-1200ms  
kernelTime is ~20ms  

I'm re-running inference on the same input twice and looking at second run to allow for warmup time of `WebGL` backend (shader compile, etc.)  

I even tried `tf.enableDebugMode()` and I still don't see anything that gets even close to overall wall time  

And I have no idea where is time spent?  

Model in question: <https://github.com/vladmandic/nanodet>

Environment: TFJS 3.3.0 on Ubuntu 20.10 and Chrome 89

**Comment by pyu10055** [comment]
@vladmandic Have you tried with our [benchmark tool](https://tensorflow.github.io/tfjs/e2e/benchmarks/local-benchmark/index.html), I tried the file in model-tfjs-graph-m/ directory, it runs around 79ms on my mac pro 2018, but the profile time is around 1sec.
The profile time is different from the real inference time, since it is wait for GL time query for each kernel.

**Comment by vladmandic** [comment]
haven't used that tool before, will try.

what confuses me that i have two variations of the model `model-tfjs-graph-m` and `model-tfjs-graph-g`  
they both have comparable execution time in `nodejs`, but, *g-variation* executes 10x faster than *m-variation* using `webgl` backend - and i can't figure out why the *m-variation* would be so slow in browser and totally fine otherwise.

**Comment by vladmandic** [comment]
Using benchmark tool, setting backend to `webgl` and leaving everything else at default values:

Model URL: <https://raw.githubusercontent.com/vladmandic/nanodet/main/model-tfjs-graph-m/nanodet.json>
Result: Subsequent average(5 runs)	114.8 ms

Model URL: <https://raw.githubusercontent.com/vladmandic/nanodet/main/model-tfjs-graph-g/nanodet.json>
Result: Subsequent average(5 runs)	65.6 ms

So difference is ~2x which is expected and not even close to ~10x that is seen in reality  
I still don't know why the model inference for m-variation is that slow - about 5x slower than expected

**Comment by vladmandic** [comment]
Another model with same symptoms: <https://github.com/vladmandic/efficientpose>

Looking at simplest "I-Lite" variation:
<https://raw.githubusercontent.com/vladmandic/efficientpose/main/model-tfjs-graph-i-lite/efficientpose.json>

Wall times (measured in JS exactly before and after call to `executeAsync()`):
- **NodeJS**: ~150 ms
- **WASM**: ~350 ms
- **WebGL**: ~500 ms

And what does benchmark tool report?
- **WASM**: subsequent average = 300 ms  
  So only slight overhead between benchmark and actual wall time - completely within reason
- **WebGL**: subsequent average = 30 ms  
  So where does the 15x difference between benchmark and wall time come from???

There is something really wrong with WebGL when executing specific models

**Comment by pyu10055** [comment]
@vladmandic The profile() method has much more overhead in WebGL comparing to WASM. It aims to measure kernel speed, no to measure inference speed. It adds GL time query at each kernel execution. The average inference time in benchmark is the more accurate measurement for deployment.

**Comment by vladmandic** [comment]
@pyu10055 I only brought up `profile()` method because I was trying to figure out why some models are  
5-10x slower for inference using `WebGL` backend than using NodeJS `tensorflow` backend (that should definitely not be the case, especially how low-profile models are) - that is the key question  

And when possible, I've tried using same models through other frameworks - for example, inference time using NodeJS with `tensorflow` backend is comparable to TF2 when used directly via Python or even with PyTorch - so no issues other than `WebGL`!

But given they profile shows total kernel time 15x lower than inference time, how can I get to bottom of this?

I've posted here two very simple models, `NanoDet` and `EfficientDet` that exhibit that behavior

**Comment by pyu10055** [comment]
@vladmandic Can you clarify on how are you benchmarking WebGL that give you 15x slowness?
For example what is the total kernel time you are referring to?
and how are you measuring the inference time?

**Comment by vladmandic** [comment]
inference time is simple:

in browser:
```js
const t0 = performance.now();
const res = await model.executeAsync(input));
const t1 = performance.now();
const inferenceTime = Math.round(t1 - t0); // elapsed time in ms
```

and in nodejs:
```js
const t0 = process.hrtime.bigint();
const res = await model.executeAsync(input)); // also can be model.predict(input);
const t1 = process.hrtime.bigint();
const inferenceTime = Math.round((t1 - t0) / 1000 /1000); // to convert to ms
```

and i'm only measuring subsequent runs, skipping first run so webgl has time to warmup (compile&upload shaders)

for most models, webgl is fast and it's the best option (since node-gpu doesn't work due to obsolete cuda dependency from tf1).
however, for some models (i've provided examples), webgl is 10x slower than nodejs.
but running `profile()` shows nothing useful.

**Comment by danwexler** [comment]
>  (since node-gpu doesn't work due to obsolete cuda dependency from tf1).

Note that tfjs-node-gpu [will soon be updated to use TF 2](https://github.com/tensorflow/tfjs/pull/4810), and it can be used with TF 1.15.

**Comment by vladmandic** [comment]
@DanielRuf yup, i'm monitoring that. still need to fix whatever is wrong with `webgl`

**Comment by pyu10055** [comment]
@vladmandic based on my test the inference time on webgl (79ms) for nandet is faster than node.js (100ms).
I am still confused about your claim that webgl is 10x slower than nodejs.
To reconciling the conceptual difference, can you provide the exact inference time you getting and how are you measure that for a single model? thanks

**Comment by vladmandic** [comment]
i've re-converted the models while changing some resource types  
it seems there was some serious clipping in the converted model (original implementation heavily uses `int64`)  
which was causing trouble for `webgl` inference as the model was branching differently and doing far more work  

right now, results are as expected - `webgl` is faster than `nodejs` in pretty much all cases  

sorry for the false leads and time wasted, this took me a while to figure out and outputs of `profile()` were throwing me off

**Comment by rohanmuplara** [comment]
@vladmandic any comments on this issue https://github.com/tensorflow/tfjs/issues/4907. Do you notice any problems between benchmark tool and own profiling for first run?

**Comment by OlivierMns** [comment]
> i've re-converted the models while changing some resource types it seems there was some serious clipping in the converted model (original implementation heavily uses `int64`) which was causing trouble for `webgl` inference as the model was branching differently and doing far more work

Hey @vladmandic , I think I am facing similar issues. I'm trying to do some inference with an EfficientNet model, and I get slower inference time with the webgl backend. Could you please give some details on how you re-converted the models to make it work ? 
Thank you!

**Comment by vladmandic** [comment]
@OlivierMns I never stated I saw the problem with EfficientNet, I spoke about EfficientDet. But in general, issue in my case was clipping due to quantization - try unquantized model first. Also, EfficientNet is notoriously slow for warmup, I suggest to enable WebGL uniforms which speeds up warmup 2-4x (no difference on actual inference)

> tf.ENV.set('WEBGL_USE_SHAPES_UNIFORMS', true);
